cli/src/commands/fix.rs

```diff
@@ -48,106 +69,253 @@ pub(crate) fn cmd_fix(
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
     let root_commits: Vec<Commit> = workspace_command
-        .parse_union_revsets(&args.sources)?
+        .parse_union_revsets(if args.sources.is_empty() {
+            &[RevisionArg::AT]
+        } else {
+            &args.sources
+        })?
         .evaluate_to_commits()?
         .try_collect()?;
     workspace_command.check_rewritable(root_commits.iter().ids())?;
 
-    let mut tx = workspace_command.start_transaction();
+    let mut tx: WorkspaceCommandTransaction = workspace_command.start_transaction();
 
-    // Collect all FileIds we're going to format and which commits they appear in
+    // Collect all of the unique tool inputs we're going to use. Tools should be
+    // deterministic, and should not consider outside information, so it is safe to
+    // deduplicate inputs that correspond to multiple files or commits. This is
+    // typically more efficient, but it does prevent certain use cases like
+    // providing commit IDs as inputs to be inserted into files. We also need to
+    // record the mapping between tool inputs and paths/commits, to efficiently
+    // rewrite the commits later.
+    //
+    // If a path is being fixed in a particular commit, it must also be fixed in all
+    // that commit's descendants. We do this as a way of propagating changes,
+    // under the assumption that it is more useful than performing a rebase and
+    // risking merge conflicts. In the case of code formatters, rebasing wouldn't
+    // reliably produce well formatted code anyway. Deduplicating inputs helps
+    // to prevent quadratic growth in the number of tool executions required for
+    // doing this in long chains of commits with disjoint sets of modified files.
     let commits: Vec<_> = RevsetExpression::commits(root_commits.iter().ids().cloned().collect())
         .descendants()
         .evaluate_programmatic(tx.base_repo().as_ref())?
         .iter()
         .commits(tx.repo().store())
         .try_collect()?;
-    let mut file_ids = HashSet::new();
-    let mut commit_paths: HashMap<CommitId, Vec<RepoPathBuf>> = HashMap::new();
+    let mut all_tool_inputs: HashSet<ToolInput> = HashSet::new();
+    let mut commit_tool_inputs: HashMap<CommitId, HashMap<RepoPathBuf, Vec<ToolInput>>> =
+        HashMap::new();
     for commit in commits.iter().rev() {
-        let mut paths = vec![];
-        // Paths modified in parent commits in the set should also be updated in this
-        // commit
-        for parent_id in commit.parent_ids() {
-            if let Some(parent_paths) = commit_paths.get(parent_id) {
-                paths.extend_from_slice(parent_paths);
-            }
-        }
-        let parent_tree = commit.parent_tree(tx.repo())?;
+        // Fix paths that are different from all parents. This means that files in merge
+        // commits are only fixed if they have evil changes, or if a parent was fixed.
```

Comment:
This doesn't seem to actually compare with each parent individually (and I don't think it should). Instead, it merges the parent trees (`commit.parent_tree()` does that) and diffs that tree against the current commit. That's how jj always defines the changes in a commit. I think "Fix paths that were changed in this commit" is enough (and more correct) here.

Reply:
I was trying to explain how I understood the behavior of diffs in the conflict and multi-parent cases, but I'm fine with the rewording. Do you think the second sentence is accurate?

I did experiment with diffing against each parent separately, because it's more similar to what `hg fix` does, but it seems complicated and unnecessary for now. This issue will come up again when adding support for changed line ranges, though; computing them against the parent(s) adds data dependencies that prevent parallelism. The line ranges to be formatted must be determined in part by either the subprocesses or diffs against ancestors (called "baserevs" in `hg fix`).

Reply:
> Do you think the second sentence is accurate?

Yes, but that's how jj always works (`jj diff -r`, `jj rebase`, `jj squash --from`, etc.) so I don't think we need to mention it. I also don't really like the term "evil merge" because I think they're called evil only because other VCS are not good at working with them so it's easy to sneak in changes. That's not true for jj.

Reply:
I'll remove it, it's not load bearing :)